src/main/java/com/cst438/service/RegistrarServiceProxy.java

```diff
@@ -0,0 +1,199 @@
+import com.cst438.dto.EnrollmentDTO;
+import com.cst438.dto.SectionDTO;
+import com.cst438.dto.UserDTO;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.amqp.core.Queue;
+import org.springframework.amqp.rabbit.annotation.RabbitListener;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.web.server.ResponseStatusException;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+@Service
+public class RegistrarServiceProxy {
+
+    @Autowired
+    private CourseRepository courseRepository;
+
+    @Autowired
+    private SectionRepository sectionRepository;
+
+    @Autowired
+    private TermRepository termRepository;
+
+    @Autowired
+    private GradeRepository gradeRepository;
+
+    @Autowired
+    private EnrollmentRepository enrollmentRepository;
+
+    @Autowired
+    private UserRepository userRepository;
+
+    BCryptPasswordEncoder encoder = new BCryptPasswordEncoder();
+
+    Queue registrarServiceQueue = new Queue("registrar_service", true);
+
+    @Bean
+    public Queue createQueue() {
+        return new Queue("gradebook_service", true);
+    }
+
+    @Autowired
+    RabbitTemplate rabbitTemplate;
+
+    public void updateFinalGrade(EnrollmentDTO e) {
+        sendMessage("updateFinalGrade " + asJsonString(e));
+    }
+
+    @RabbitListener(queues = "gradebook_service")
+    public void receiveFromRegistrar(String message)  {
+        try {
+            String[] parts = message.split(" ", 1);
+            String action = parts[0];
+            if (action.equals("addCourse")) {
+                CourseDTO dto = fromJsonString(parts[1], CourseDTO.class);
+                Course c = new Course();
+                c.setCourseId(dto.courseId());
+                c.setTitle(dto.title());
+                c.setCredits(dto.credits());
+                courseRepository.save(c);
+            } else if (action.equals("deleteCourse")) {
+                courseRepository.deleteById(parts[1]);
+            } else if (action.equals("updateCourse")) {
+                CourseDTO dto = fromJsonString(parts[1], CourseDTO.class);
+                Course c = courseRepository.findById(dto.courseId()).orElse(null);
+                c.setTitle(dto.title());
+                c.setCredits(dto.credits());
+                courseRepository.save(c);
+            } else if (action.equals("addSection")) {
+                SectionDTO dto = fromJsonString(parts[1], SectionDTO.class);
+                Course course = courseRepository.findById(dto.courseId()).orElse(null);
+                Section s = new Section();
+                s.setCourse(course);
+                Term term = termRepository.findByYearAndSemester(dto.year(), dto.semester());
```

Comment:
Check if the term is null before setting it to the section.
Example:
if (term == null) {
    throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Term not found for year " + dto.year() + " and semester " + dto.semester());
}

Reply:
Addressed this.